doc/services/rtio/index.rst

```diff
@@ -0,0 +1,417 @@
+
+1. API is DMA and interrupt friendly.
+2. No buffer copying
+3. No callbacks
+4. Blocking or non-blocking operation
+
+Problem
+*******
+
+An application wishing to do complex DMA or interrupt driven operations today
+in Zephyr requires direct knowledge of the hardware and how it works.
+
+This means doing complex audio, video, or sensor streaming requires direct
+hardware knowledge or leaky abstractions over DMA controllers. Neither is ideal.
+
+To enable asynchronous operations, especially with DMA, a description of what
+to do rather than direct operations through C and callbacks is needed. Enabling
+DMA features such as channels with priority, and sequences of transfers requires
+more than a simple list of descriptions.
+
+Using DMA and/or interrupt driven I/O shouldn't dictate whether or not the
+call is blocking or not.
+
+Inspiration, introducing io_uring
+*********************************
+
+It's better not to reinvent the wheel (or ring in this case) and io_uring as an
+API from the Linux kernel provides a winning model. In io_uring there are two
+lock-free ring buffers acting as queues shared between the kernel and a userland
+application. One queue for submission entries which may be chained and flushed to
+create concurrent sequential requests. A second queue for completion events.
+Only a single syscall is actually required to execute many operations, the
+io_uring_submit call. This call may block the caller when a number of
+operations to wait on is given.
+
+This model maps well to DMA and interrupt driven transfers. A request to do a
+sequence of operations in an asynchronous way directly relates
+to the way hardware typically works with interrupt driven state machines
+potentially involving multiple peripheral IPs like bus and DMA controllers.
+
+Submission Queue and Chaining
+*****************************
+
+The submission queue (sq), is the description of the operations
+to perform in concurrent chains.
+
+For example imagine a typical SPI transfer where you wish to write a
+register address to then read from. So the sequence of operations might be...
+
+   1. Chip Select
+   2. Clock Enable
+   3. Write register address into SPI transmit register
+   4. Read from the SPI receive register into a buffer
+   5. Disable clock
+   6. Disable Chip Select
+
+If anything in this chain of operations fails give up. Some of those operations
+can be embodied in a device abstraction that understand a read or write
+implicitly means setup the clock and chip select. The transactional nature of
+the request also needs to be embodied in some manner. Of the operations above
+perhaps the read could be done using DMA as its large enough make sense. That
+requires an understanding of how to setup the device's particular DMA to do so.
+
+The above sequence of operations is embodied in RTIO as chain of
+submission queue entries (sqe). Chaining is done by setting a bitflag in
+an sqe to signify the next sqe must wait on the current one.
```

Comment:
Chaining is meant to say the current SQE MUST complete before the next one is started. The wording here is wrong clearly.

Chains can be thought of as independent threads of execution. You wouldn't expect two C statements to run in parallel, the mental model is one is done the next statement is started.

The submission queue while being a queue itself, is a description of numerous concurrent chains of operations that can be done, ideally driven entirely by hardware and hardware events. But they could be larger operations that map to potentially many hardware events and operations such as a transfer from a sensor or audio codec.

The io_uring guide probably best describes this, direct quote from
[io_uring.pdf](https://kernel.dk/io_uring.pdf)

> The chain of linked sqes continue as long as IOSQE_IO_LINK is set in the flags ﬁeld. Hence the chain is deﬁned as
> starting with the ﬁrst sqe that has IOSQE_IO_LINK set, and ends with the ﬁrst subsequent sqe that does not have it set.
> Arbitrarily long chains are supported.
> The chains execute independently of other sqes in the submission ring. Chains are independent execution units, and
> multiple chains can execute and complete in parallel to each other. This includes sqes that are not part of any chain.

Now is this the right model for RTIO and an RTOS is another question altogether! That's what I'm attempting to answer for myself and so far I think makes sense.

Reply:
I could see having input from multiple sensors on different busses that send their data to another software sensor that combines them and sends output to the original caller. This could include async operations happening in parallel to the physical sensors but the software sensor may not be able to do anything until all of the data from all of the sensors comes in... Thus putting up a wall to wait for all of the async chains to join back would be a good way to send this data on to the software sensor.

Reply:
I can definitely see that! That really is creating a processing pipe. There's potential other flows like that even with something like a soft core on an fpga with a hardware accelerator I can imagine as well, I agree it needs a full pipe wait mode of some kind for that sort of thing.